### DISCUSSION TITLE
Cost of conversion from json to ordered_json

### DISCUSSION THREAD

**Discussion by pozemka** [nlohmann/json#3335] [Q&A] [answered]
**See comments to accepted answer**

Hello.

I have a function that takes some JSON and processes it. Lets say`void processJson(const nlohmann::json& json_data);`

In some cases however I call it with ordered_json argument:
```
nlohmann::ordered_json json_data = getOrderedJson();
processJson(json_data );
```

Question is: How expensive is conversion from ordere_json to json and vice versa? Should I overload my function or having one is just fine.

I understand, best way to know is to test myself, but for now I just don't have enough data for really good test.

**See comments to accepted answer**

**Comment by gregmarr** [comment, marked as answer]
It would be exactly the same as if your function was `void processJson(nlohmann::json json_data);`.  That is, it makes a deep copy of the object.  So it depends on the size of your object.  You're probably better off doing

    void processJson(const nlohmann::basic_json<> &json_data) {
    }

**Reply by pozemka** [reply]
Thank you very much! I will try this approach.

Didn't knew that `json` and `basic_json<>` act different in this case, since first is basically the second:

`using json = basic_json<>;`

**Reply by gregmarr** [reply]
Hm, you may have to actually spell out the template then.  I forgot that all the `basic_json<>` template parameters were defaulted.

**Reply by gregmarr** [reply]
This works, but there's no requirement that T actually be `basic_json` so it may cause issues if you call it with the wrong thing.  That may be perfectly okay in the way that you're using it.   Otherwise, you'll need to do the full template specialization.

    template<typename T> void processJson(T const &json_data) {
    }

**Reply by pozemka** [reply]
For now I decided to use templates with additional checks to avoid mixing `json` types with `string` type which are both can implicitly convert to each other:
```cpp
template <typename TJ, std::enable_if_t<std::is_same_v<TJ, nlohmann::json> || std::is_same_v<TJ, nlohmann::ordered_json> ,bool> = false>
size_t processJSON(const TJ& json, const std::string& name)
{
...
}
```
Not the most elegant solution but it seems the best one for now.

**Comment by gregmarr** [comment]
You could do this except that the macros are undefined at the end of the include and you can't include them again.  These might be something useful to have in the `_fwd` header.

    NLOHMANN_BASIC_JSON_TPL_DECLARATION
    void processJson(const NLOHMANN_BASIC_JSON_TPL& j)
